### PR TITLE
First pass - high res download

### PR DESCRIPTION
**Pull request by jamesmorgan**
* [x] Use metamask to sign message
* [x] post to backend for verification
* [x] handle metamask account not unlocked
* [x] handle failure
* [x] Present user with high-res one success
* [x] finish backend

### CODE REVIEW

File: contracts/KnownOriginDigitalAssetBatchMinter.sol [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+* BE ORIGINAL. BUY ORIGINAL.
+*
+*/
+contract KnownOriginDigitalAssetBatchMinter {
+
+  address public owner;
+  address public curatorAccount;
+
+  modifier onlyKnownOrigin() {
+    require(msg.sender == curatorAccount || msg.sender == developerAccount);
```
Comment (andygray) [inline contracts/KnownOriginDigitalAssetBatchMinter.sol]: no ref to `developerAccount` - should be owner?

File: contracts/KodaMintableInterface.sol [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+interface KodaMintableInterface {
+  /**
+   * @dev Mint a new KODA token
+   * @dev Reverts if not called by management
+   * @param _tokenURI the IPFS or equivalent hash
+   * @param _edition the identifier of the edition - leading 3 bytes are the artist code, trailing 3 bytes are the asset type
+   * @param _priceInWei the price of the KODA token
+   * @param _auctionStartDate the date when the token is available for sale
+   */
+  function mint(string _tokenURI, bytes16 _edition, uint256 _priceInWei, uint32 _auctionStartDate, address _artistAccount) external;
```
Comment (andygray) [inline contracts/KodaMintableInterface.sol]: Nice. I've seen `IKodaMintable` used but happy either way

File: src/store/index.js [hunk 3]
```diff
@@ -67,7 +68,8 @@ const store = new Vuex.Store({
     findNextAssetToPurchase: (state, getters) => (edition) => {
       let editions = getters.assetsForEdition(edition.edition);
       return _.chain(editions)
-        .orderBy('editionNumber')
+      // find cheapest next edition (some editions can have different prices)
```
Comment (andygray) [inline src/store/index.js]: we have not used this tactic yet?

File: src/store/index.js [hunk 4]
```diff
@@ -268,14 +282,14 @@ const store = new Vuex.Store({
...
-        [tokenId]: {tokenId: tokenId.toString(10), buyer, state: 'PURCHASE_FAILED'}
+        [tokenId]: {tokenId: tokenId.toString(), buyer, state: 'PURCHASE_FAILED'}
       };
     },
     [mutations.PURCHASE_SUCCESSFUL](state, {tokenId, buyer}) {
       state.purchaseState = {
         ...state.purchaseState,
         [tokenId]: {
-          tokenId: tokenId.toString(10),
+          tokenId: tokenId.toString(),
```
Comment (andygray) [inline src/store/index.js]: why drop the `10` bit - guess doesn't matter as tokenId unlikely to get too big where `e15` format etc is used

File: src/store/index.js [hunk 5]
```diff
@@ -804,16 +833,80 @@ const store = new Vuex.Store({
...
+      commit(mutations.HIGH_RES_DOWNLOAD_TRIGGERED, {tokenId: asset.id});
+
+      const requestHighResDownload = ({originalMessage, signedMessage}) => {
+        return state.web3.eth.net.getId()
+          .then((networkId) => {
+
+            const highResConfig = {
+              local: "http://localhost:5000/known-origin-io/us-central1/highResDownload",
+              beta: "https://us-central1-beta-known-origin-io.cloudfunctions.net/highResDownload",
+              live: "https://us-central1-known-origin-io.cloudfunctions.net/highResDownload"
```
Comment (andygray) [inline src/store/index.js]: any risk exposing this address?

File: contracts/KnownOriginDigitalAssetBatchMinter.sol (hunk 1 above)
Comment (jamesmorgan) [reply contracts/KnownOriginDigitalAssetBatchMinter.sol]: ye will change to `owner` - this is untested/unused so not to worry

File: contracts/KodaMintableInterface.sol (hunk 2 above)
Comment (jamesmorgan) [reply contracts/KodaMintableInterface.sol]: better name 👍

File: src/components/ui-controls/HighResDownload.vue [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        'isHighResDownloadSuccess',
+        'isHighResDownloadFailed',
+        'hasHighResDownloadedLink'
+      ]),
+    },
+    methods: {
+      verifyPurchase: function () {
+        this.$store.dispatch(actions.HIGH_RES_DOWNLOAD, this.asset);
+      },
+      accountOwnsAsset: function () {
```
Comment (jamesmorgan) [inline src/components/ui-controls/HighResDownload.vue]: Change this to `shouldDisplayHighRes` and remove additional `&` in html

File: src/store/index.js (hunk 5 above)
Comment (jamesmorgan) [reply src/store/index.js]: Not sure what the alternative is, the web app needs to know the API?

File: src/components/ui-controls/HighResDownload.vue (hunk 6 above)
Comment (andygray) [reply src/components/ui-controls/HighResDownload.vue]: 👍

File: src/store/index.js (hunk 3 above)
Comment (jamesmorgan) [reply src/store/index.js]: I think some editions have different prices

**Pull request merged**